src/sg/aca_security_group_ovs.cpp

```diff
@@ -0,0 +1,610 @@
+// Copyright 2019 The Alcor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include<sstream>
+#include "aca_log.h"
+#include "aca_security_group_ovs.h"
+#include "aca_security_group_manager.h"
+#include "aca_ovs_control.h"
+
+using namespace aca_ovs_control;
+
+namespace aca_security_group {
+
+static const char *ct_state[] = {
+	"+est-rel-rpl",
+	"+new-est"
+};
+
+Aca_Security_Group_Ovs &Aca_Security_Group_Ovs::get_instance()
+{
+    // It is instantiated on first use.
+    // allocated instance is destroyed when program exits.
+    static Aca_Security_Group_Ovs instance;
+    return instance;
+}
+
+int Aca_Security_Group_Ovs::get_vlan_by_segment_id(const int segment_id)
+{
+    return segment_id;
+}
+
+void Aca_Security_Group_Ovs::init_port_egress_flows(Aca_Port &port) 
+{
+    ACA_OVS_Control &controller = ACA_OVS_Control::get_instance();
+    char flow[128] = {0};
+
+    string port_id = port.get_id();
+    string vpc_id = port.get_vpc_id();
+    const char *mac_address = port.get_mac_address().data();
+    int ofport = 12;
+    int patch_ofport = 13;
+
+    int vni = 10;
+    int local_vlan = get_vlan_by_segment_id(vni);
+
+
+    // DHCP discovery
+    sprintf(flow, "table=%d,priority=80,in_port=%d,reg%d=%d,dl_type=0x%04x,nw_proto=%d,"
+        "tp_src=68,tp_dst=67,actions=resubmit(,%d)", BASE_EGRESS_TABLE, ofport, REG_PORT, 
+        ofport, ETHERTYPE_IP, PROTO_NUM_UDP, ACCEPT_OR_INGRESS_TABLE);
+    controller.add_flow(BR_INT, flow);
+
+    // Drop DHCP response from vm
+    sprintf(flow, "table=%d,priority=70,in_port=%d,reg%d=%d,dl_type=0x%04x,nw_proto=%d,"
+        "tp_src=67,tp_dst=68,actions=resubmit(,%d)", BASE_EGRESS_TABLE, ofport, REG_PORT, 
+        ofport, ETHERTYPE_IP, PROTO_NUM_UDP, DROPPED_TRAFFIC_TABLE);
+    controller.add_flow(BR_INT, flow);
+
+    // Drop all remaining egress connections
+    sprintf(flow, "table=%d,priority=10,in_port=%d,reg%d=%d,actions=ct_clear,resubmit(,%d)", 
+        BASE_EGRESS_TABLE, ofport, REG_PORT, ofport, DROPPED_TRAFFIC_TABLE);
+    controller.add_flow(BR_INT, flow);
```

Comment:
add_flow can fail and returns and error code.

Do you want to add the flows as bundle so that it is "all or nothing"

@cj-chung - FYI.